Add a separate heartbeatTimeout configuration.

petermattis opened the pull request:
Tests were previously computing heartbeat timeouts of 20ms which would
sometimes expire on very slow/overloaded machines. With a separate
heartbeatTimeout configuration, only tests which need a low
heartbeatTimeout bother to change it.

De-flake a few other tests by increasing timeouts or removing them
entirely on relying on the test timeout.

Fixes #3734.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3790)

<!-- Reviewable:end -->

petermattis committed:
Add a separate heartbeatTimeout configuration.

Tests were previously computing heartbeat timeouts of 20ms which would
sometimes expire on very slow/overloaded machines. With a separate
heartbeatTimeout configuration, only tests which need a low
heartbeatTimeout bother to change it.

De-flake a few other tests by increasing timeouts or removing them
entirely on relying on the test timeout.

Fixes #3734.

petermattis commented on rpc/send_test.go:
```diff
@@ -241,28 +243,27 @@ func TestClientNotReady(t *testing.T) {
 	go func() {
 		if _, err := sendPing(opts, []net.Addr{ln.Addr()}, nodeContext); err == nil {
-			t.Fatalf("expected error when client is closed")
+			c <- fmt.Errorf("expected error when client is closed")
```
It's a data race to call `t.Fatalf` from a goroutine other than the one that started the test.

petermattis commented on rpc/client.go:
```diff
@@ -306,6 +324,9 @@ func (c *Client) runHeartbeat(retryOpts retry.Options) {
 			if err = c.heartbeat(retryOpts.Closer); err != nil {
 				setUnhealthy()
 				log.Warning(err)
+				if c.maybeClose(retryOpts.Closer) {
```
I'm not sure if this was necessary. In one test failure the client noticed it was closed (`err == errClosed`) but then hit the retry loop again before actually closing. That seemed wrong.

bdarnell commented on rpc/client.go:
```diff
@@ -131,6 +133,15 @@ func NewClient(addr net.Addr, context *Context) *Client {
+		c.heartbeatInterval = defaultHeartbeatInterval
+	}
+	if c.heartbeatTimeout == 0 {
+		c.heartbeatTimeout = defaultHeartbeatTimeout
```
I think `c.heartbeatTimeout = 2 * c.heartbeatInterval` is still a better default than a hard-coded 6s (which basically forces anyone who sets one to set the other). The test should set an explicit `heartbeatTimeout` instead of relying on the default computed one.

tamird commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[rpc/send_test.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3790#-K7rWF_01XBsk1ODPVCJ-r1-249:-K7rZIE1hFzmQRDkcriK:1933098195) ([raw file](https://github.com/cockroachdb/cockroach/blob/20979a6962e4cd4f5431d0ccb043d38ef1f2d4cf/rpc/send_test.go#L249)):*
nit: `util.Errorf` since the line number Go will report will point to the `Fatal` call below.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3790)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

petermattis commented:
Review status: 3 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[rpc/client.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3790#-K7rWF_01XBsk1ODPVCE-r1-144:-K7rpssAItuDQg8YBL0u:216577204) ([raw file](https://github.com/cockroachdb/cockroach/blob/20979a6962e4cd4f5431d0ccb043d38ef1f2d4cf/rpc/client.go#L144)):*
I've changed this and fixed up `newNodeTestContext` to set the default heartbeat timeout for tests to 6 secs.

---

*[rpc/send_test.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3790#-K7rWF_01XBsk1ODPVCJ-r1-249:-K7rpkxycrRgcwX5OXdD:1996155710) ([raw file](https://github.com/cockroachdb/cockroach/blob/20979a6962e4cd4f5431d0ccb043d38ef1f2d4cf/rpc/send_test.go#L249)):*
This is only used in a test so it doesn't really matter, but done anyways.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3790)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3790)*
<!-- Sent from Reviewable.io -->
